activesupport/lib/active_support/testing/assertions.rb

```diff
@@ -261,6 +261,34 @@ def assert_no_changes(expression, message = nil, from: UNTRACKED, &block)
         retval
       end
 
+      # Assertion that an active model doesn't have an error on a field
+      #
+      #   assert_no_error :name, :blank, :user
+      def assert_no_error(attribute, type, obj, msg = nil)
+        raise ArgumentError.new("#{obj.inspect} does not respond to #validate") unless obj.respond_to?(:validate)
+
+        obj.validate
+        msg = message(msg) {
+          data = [attribute, type]
+          "Expected %s to not be %s" % data
+        }
+        assert_empty obj.errors.where(attribute, type), msg
+      end
+
+      # Assertion that an active model has a specific error on a field
+      #
+      #   assert_error :name, :blank, :user
+      def assert_error(attribute, type, obj, msg = nil)
+        raise ArgumentError.new("#{obj.inspect} does not respond to #validate") unless obj.respond_to?(:validate)
+
+        obj.validate
+        msg = message(msg) {
+          data = [attribute, type]
+          "Expected %s to be %s" % data
+        }
+        refute_empty obj.errors.where(attribute, type), msg
+      end
```

Comment:
To share some prior art in this space, [thoughtbot/shoulda-matchers](https://github.com/thoughtbot/shoulda-matchers?tab=readme-ov-file#matchers) might be of interest. It is a [14 year old](https://github.com/thoughtbot/shoulda-matchers/commit/ecb01a209d08ae769f1c2eafb14680b1d2a73e98) project that provides Active Model validation assertions (amongst other things).

Like this proposal, it also invokes `#validate` on the object directly. Unlike this proposal, it names its matches in a way that communicates that it's invoking `#validate` (for example, `validates_presence_of`, `validates_uniqueness_of`, etc.).

Callers of an `assert_no_error` assertions might be surprised to learn that it's also invoking `#validate`, since `#validate` clears any errors present on the object before re-validating. Depending on the details of the test itself, that kind of reseting might un-do and remove errors added during that test.

As a caller, I might expect `assert_no_error` to _only_ make assertions about the errors available by the subject under test's [#errors](https://edgeapi.rubyonrails.org/classes/ActiveModel/Validations.html#method-i-errors) method.

If this proposal is deemed viable, what do you think about introducing both `assert_valid` and `assert_not_valid` assertions that invoke `assert_no_error` and `assert_error` (respectively)? That way, callers can benefit from the convenience of `assert_valid` or `assert_not_valid` while also being able to drop down to the finer-grained `assert_error` or `assert_no_error`.

I'm imagining something like:

```ruby
      # Assertion that an active model doesn't have an error on a field
      #
      #   assert_no_error :name, :blank, :user
      def assert_no_error(attribute, type, obj, msg = nil)
        msg = message(msg) {
          data = [attribute, type]
          "Expected %s to not be %s" % data
        }
        assert_empty obj.errors.where(attribute, type), msg
      end
      
      # Assertion that an active model has a specific error on a field
      #
      #   assert_error :name, :blank, :user
      def assert_error(attribute, type, obj, msg = nil)
        msg = message(msg) {
          data = [attribute, type]
          "Expected %s to be %s" % data
        }
        refute_empty obj.errors.where(attribute, type), msg
      end
      
      def assert_not_valid(attribute, type, obj, msg = nil)
        raise ArgumentError.new("#{obj.inspect} does not respond to #validate") unless obj.respond_to?(:validate)

        obj.validate
        assert_error(attribute, type, object, msg)
      end

      def assert_valid(attribute, type, object, msg = nil)
        raise ArgumentError.new("#{obj.inspect} does not respond to #validate") unless obj.respond_to?(:validate)

        obj.validate
        assert_no_error(attribute, type, object, msg)
      end
```

Reply:
Are there plans to support [Active Model Validation Contexts](https://guides.rubyonrails.org/active_record_validations.html#on)?

When invoking the version of `assert_no_error` that *does not call #validate*, it's up to the test code that precedes the assertion to validate with the proper context.

When invoking the `assert_valid` and `assert_not_valid` variations, there isn't a way to forward along any contextual information (like an `:on` option). This is also true of the currently proposed `assert_error` and `assert_no_error` definitions that call `#validate` directly.

Reply:
> Callers of an assert_no_error assertions might be surprised to learn that it's also invoking #validate
Agreed, thanks for the feedback. I will go ahead and make those changes.

Regarding 
> Are there plans to support Active Model Validation Contexts?

This can be another type of assertion. However, the current change aims to focus on a model that went through the process of doing its validations. 


PS: This is my first contribution ever to Rails and if it's better to ship a more defined feature then I am onboard with adding more stuff. LMK what's the approach generally taken. I am thinking this in terms of small deliverable.